USAGE.md

```diff
@@ -0,0 +1,125 @@
+Example Usage
+=============
+
+Not entirely aligned with current `allocator_common.h`, just wanted to try
+and capture some thoughts about how the usage of this API might roughly look.
+
+Note that we want to take into account scenarios where some processes may not
+be permitted to open all devices.  For example a client of the display server
+running in a sandbox, etc.  And we don't want window-system specific IPC
+knowledge in "liballoc".
+
+So each `alloc_dev_t` should be exposed to the user, and logic for
+constructing the union of constraints and intersection of capabilities
+should be decoupled from `alloc_dev_t`.  And of course, constraints/
+capabilities should be easily serializable.
+
+To keep things simple, a hypothetical example with a camera + display, in
+single process but with some comments about where things would be serialized
+in a multi-process scenario.
+
+All error checking omitted for brevity.
+
+Open Devices
+------------
+
+Generally an `alloc_dev_t` represents a single device file or API.  In cases
+where the underlying device file is not exposed (such as OpenMAX or OpenCL,
+or OpenGL when not running on top of gbm) an extension would be required to
+retrieve the `alloc_dev_t` from that particular API.
+
+```
+alloc_dev_t *display = alloc_dev_create(open("/dev/dri/card0", O_RDWR));
+alloc_dev_t *camera  = alloc_dev_create(open("/dev/video0", O_RDWR));
```

Comment:
I had assumed the allocator might include some generic backends, just like the kernel has in-tree drivers (and err, one or two out of tree ones).  One of those might be a small V4L library that handled interfacing with the V4L device nodes.  Ideally we'd set it up such that these base libraries could be pretty small.

I had imagined there would be two layers of device objects: IDs (I.e., the UUID/LUID) and instances (alloc_dev_t).  Perhaps you could create an alloc_dev_t directly from a UUID, or from a (UUID, fd) pair for cases where you get an FD over the wire, or directly from an FD, bypassing the UUID mechanism.  DRM for example, has seemingly never-ending APIs for opening the right devnode.  That seems like a somewhat excessive example, but having more than one would probably be reasonable.

The reason I like having a UUID/LUID in the API is they serialize quite easily, even between containers.  If you identify devices only using an alloc_dev_t (presumably a pointer to some local structure), or a file descriptor or filename, it's harder to send between processes.  Maybe in your container all the devices are in /container/dev instead of /dev (This is probably a stupid example, but I think it's still a reasonable concern), or maybe you want to figure out how to correlate between some file descriptor your compositor passed you and some Vulkan physical device.  If you have a UUID/LUID, and you can query that from the Vulkan physical device using a to-be-written extension, correlation is easy.